pkg/pushsync/pushsync.go

```diff
@@ -352,60 +349,68 @@ func (ps *PushSync) pushToClosest(ctx context.Context, ch swarm.Chunk, retryAllo
 			}
 			return nil, fmt.Errorf("closest peer: %w", err)
 		}
-
-		skipPeers = append(skipPeers, peer)
-
-		if ps.skipList.ShouldSkip(peer) {
-			ps.metrics.TotalSkippedPeers.Inc()
-			continue
-		}
-
 		ps.metrics.TotalSendAttempts.Inc()
 
-		go func(peer swarm.Address, ch swarm.Chunk) {
-			ctxd, canceld := context.WithTimeout(ctx, defaultTTL)
-			defer canceld()
+		ctxd, canceld := context.WithTimeout(ctx, defaultTTL)
+		defer canceld()
+		ps.logger.Debugf("pushsync: pushing chunk %s to peer %s", ch.Address().String(), peer.String())
 
-			r, attempted, err := ps.pushPeer(ctxd, peer, ch, retryAllowed)
-			// attempted is true if we get past accounting and actually attempt
-			// to send the request to the peer. If we dont get past accounting, we
-			// should not count the retry and try with a different peer again
-			if attempted {
-				allowedRetries--
-			}
-			if err != nil {
-				logger.Debugf("could not push to peer %s: %v", peer, err)
+		r, attempted, err := ps.pushPeer(ctxd, peer, ch, retryAllowed)
 
-				// if the node has warmed up AND no other closer peer has been tried
-				if ps.warmedUp() && !ps.skipList.HasChunk(ch.Address()) {
-					ps.skipList.Add(peer, ch.Address(), skipPeerExpiration)
+		// attempted is true if we get past accounting and actually attempt
+		// to send the request to the peer. If we dont get past accounting, we
+		// should not count the retry and try with a different peer again
+		if attempted {
```

Comment:
I think now that we are explicitly handling errors, wouldn't it be better to get rid of this attempted flag all together? I think it could improve the readability of the code if we handle the decrementing of allowedRetries in the switch case below.

Reply:
yes and no :) the problem is that there could still be some other error and if we haven't attempted at all we should still probably attempt a different peer. in general i agree with you, but I am not feeling confident enough that we cover all cases here at this point